### DISCUSSION TITLE
Little help on how to use pytest as multithread

### DISCUSSION THREAD

**Discussion by AlexPaiva** [seleniumbase/SeleniumBase#2241] [Q&A] [answered]
So I am using the raw driver and works perfectly, but want to multi-thread it.

Thing is this is a tkinter app packages in a .exe with pyinstaller so I am a little confused on how to use pytest this way.

I have a single function that takes arguments that generates the driver and does everything (small function, a few lines only) how could I turn it into a multithreaded function with pytest here? The number of threads being set by a thread safe global variable.

Function goes something like this:

```
from seleniumbase import Driver

def driver_gen(proxyList):
   dd = Driver(uc=True, proxy=proxyList)

Tkinter app code here that generates a separate thread and summons the driver_gen separated from the main thread already.
```

Can anyone give me a hand here?

**Comment by mdmintz** [comment, marked as answer]
As mentioned here: https://github.com/seleniumbase/SeleniumBase/discussions/2097#discussioncomment-6978972 ...

If you're looking to run multi-threaded tests in UC Mode with a proxy, here's the scoop:

Multi-threading in UC Mode is possible if you use `pytest` multi-threading provided by [pytest-xdist](https://github.com/pytest-dev/pytest-xdist).

You'll need to use command-line options for this format, eg ``--uc`` to activate UC Mode, ``-n4`` (for 4 parallel processes, etc), and ``--proxy=user:pass@host:port`` to set proxy settings.

Below is a sample run command:

```bash
pytest --uc -n4
```

> (Add ``--proxy=user:pass@host:port`` to include proxy settings.)

Here's a sample file that uses `@pytest.mark.parametrize()` to turn one test into four tests when run with ``pytest``:

```python
import pytest

@pytest.mark.parametrize("", [[]] * 4)
def test_multi_threaded(sb):
    sb.driver.get("https://nowsecure.nl/#relax")
    try:
        sb.assert_text("OH YEAH, you passed!", "h1", timeout=5.25)
        sb.post_message("Selenium wasn't detected!", duration=2.8)
        sb._print("\n Success! Website did not detect Selenium! ")
    except Exception:
        sb.fail('Selenium was detected! Try using: "pytest --uc"')
```

Here's the output when running that file with ``pytest --uc -n4``:

```bash
pytest test_multi_uc.py --uc -n4
============================ test session starts =============================
configfile: pytest.ini
4 workers [4 items]     

 Success! Website did not detect Selenium! 

 Success! Website did not detect Selenium! 
..
 Success! Website did not detect Selenium! 
.
 Success! Website did not detect Selenium! 
.
============================= 4 passed in 9.38s ==============================
```

Some websites may block you if they detect multiple simultaneous connections like that. Be careful where you go.

Note that there are different syntax formats. See: https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md

`pytest-xdist` (the multi-processing library for `pytest`) is more advanced than https://docs.python.org/3/library/concurrent.futures.html for complex multi-threading of this nature.

If you have to use multithreading without `pytest`, you'll have to implement your own thread-locking system so that no UC browser is spun up at the exact same time as another browser. Otherwise conflicts could arise when converting chromedriver into uc_driver, or when selecting a free port. (More info on that: https://github.com/seleniumbase/SeleniumBase/issues/2006#issuecomment-1787959243)

Also, if you haven't already seen the UC Mode video tutorial, then watch https://www.youtube.com/watch?v=5dMFI3e85ig, where you'll see multithreaded UC Mode examples, as well as answers to many of your UC Mode questions.

--------

You can also run `pytest` multithreaded from a subprocess. Example:

```python
subprocess.run(["pytest", "-v", "-n", "6", "test_repeat_tests.py"])
```

--------

Most example tests already have the `BaseCase.main(...)` line, where you can invoke `pytest` from a `python` call.

You can include additional args, such as `-n6`, for running tests multithreaded.

```python
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__, "-n6", "-v", "-s")
```

That would be the equivalent of running:

```python
import subprocess
import sys

if __name__ == "__main__":
    subprocess.call([sys.executable, "-m", "pytest", __file__, "-n6", "-v", "-s"])
```

Using `subprocess.call` with `sys.executable` is generally better than using `subprocess.run`. Reason: If you have multiple versions of Python installed on your system, then the same Python that invoked the script is used if you use `subprocess.call` with `sys.executable`.

**Reply by AlexPaiva** [reply]
Thank you, yes I saw that already and the video and still am confused, I know how to use pytest-xdist but via CLI, progamatically is what I have doubts about, to call from a function and not via CLI.

**Reply by mdmintz** [reply]
From the CLI is how `pytest` (with `pytest-xdist`) initiates all the multithreaded processes and manages them. That multithreaded process management is important for making sure the threads don't step on each other.

**Reply by AlexPaiva** [reply]
Ok, first of all thank you for your answer and the repository and your work is amazing, but the selected answer didn't answer exactly what I needed, please mark this as the answer as this is what I was looking for and for anyone else here is what you need to do, quite simple but I didn't initially grasp the concept easily:



Get the function/class/whatever you wanted to run and place it in a separate python file from the main file and call it for example `test_file.py`

Then on the main python file just replace the call of the function to `subprocess.run(["pytest", "-v", "-n", "3", "test_file.py"]) `and replace `3` with the number of threads you want, by using run it blocks so after you can do some garbage cleanup or something or just print 'finished'.

This will work without having to use the CLI directly. This was tested only with selenium base code not using any of pytests 'abilities' and only for the purpose of multithreading with isolation.

No need for any `@pytest.mark.parametrize("", [[]] * 3)` or anything, just place the function in a new file as if you were going to run normal python code.

**Reply by mdmintz** [reply]
Most example tests already have the `BaseCase.main(...)` line, where you can invoke `pytest` from a `python` call.

You can include additional args, such as `-n6`, for running tests multithreaded.

```python
from seleniumbase import BaseCase
BaseCase.main(__name__, __file__, "-n6", "-v", "-s")
```

That would be the equivalent of running:

```python
import subprocess
import sys

if __name__ == "__main__":
    subprocess.call([sys.executable, "-m", "pytest", __file__, "-n6", "-v", "-s"])
```

**Reply by AlexPaiva** [reply]
Yes but I was using the raw driver and my knowledge of pytest was poor alogn with wanting customizations to -n command etc... Thank you!

**Reply by AlexPaiva** [reply]
Quick question, if instead of `subprocess.run(["pytest", "-v", "-n", "2", "test.py"])` to start 2  threads I ran `subprocess.run(["pytest", "-v", "test.py"])` twice with normal python threading will it be the same or isolation not done properly or?

When I say with threading I mean something like this:

```
def call_pytest():
        subprocess.run(["pytest", "-v", "test.py"])
 
threads = []
for _ in range(2):
        thread = threading.Thread(target=call_pytest)
        threads.append(thread)
        thread.start()
        
for thread in threads:
        thread.join()
```

**Reply by mdmintz** [reply]
Unless all your threads are spun up via a single process that can control all of them and prevent overlaps, I'd suggest using the built-in `pytest` multithreading solution.

**Comment by AlexPaiva** [comment]
> Unless all your threads are spun up via a single process that can control all of them and prevent overlaps, I'd suggest using the built-in `pytest` multithreading solution.

Could you explain a little bit better what you mean exactly? I am calling from one thread all the pytests subprocesses, it should work? The issue is that I want to control each pytest test independently, if I run` -n 4` I want that when a test is done it depending on a few factors starts again or not, but with `-n 4` only when all tests are done is when I can re-start them if needed. And if I change the code inside the test itself to restart if needed, it wouldn't properly re-isolate everything correct?

**Reply by mdmintz** [reply]
I'm confused about what you're asking. `pytest` should be starting and controlling all the subprocesses. You can loop the process that calls `pytest` if necessary. I'm not sure what you mean about `control each pytest test independently`. `pytest` multithreading works via the `pytest-xdist` plugin. For more info on that, see https://github.com/pytest-dev/pytest-xdist.
